lib/types.ts

```diff
@@ -424,7 +424,13 @@ export type Opaque<Type, Token extends string> = Token extends StringLiteral<Tok
   : never;
 
 /** Easily extract the type of a given object's values */
-export type ValueOf<T> = T[keyof T];
+export type ValueOf<T> = T extends Primitive
+  ? T
+  : T extends any[]
+  ? T[number]
+  : T extends (...args: any[]) => any
```

Comment:
If possible, could you please move it to `AnyFunction`, e.g.

```ts
type AnyFunction<TArgs extends any[] = any[], TReturnType = any> = (...args: TArgs) => TReturnType;
```

and it becomes:

```suggestion
  : T extends AnyFunction
```

Reply:
Yes will add this type helper `AnyFunction` and refactor it